src/Internal/ProductAttributesLookup/Filterer.php

```diff
@@ -0,0 +1,323 @@
+<?php
+/**
+ * Filterer class file.
+ */
+
+namespace Automattic\WooCommerce\Internal\ProductAttributesLookup;
+
+defined( 'ABSPATH' ) || exit;
+
+
+/**
+ * Helper class for filtering products using the product attributes lookup table.
+ */
+class Filterer {
+
+	/**
+	 * The product attributes lookup data store to use.
+	 *
+	 * @var LookupDataStore
+	 */
+	private $data_store;
+
+	/**
+	 * The name of the product attributes lookup table.
+	 *
+	 * @var string
+	 */
+	private $lookup_table_name;
+
+	/**
+	 * Class initialization, invoked by the DI container.
+	 *
+	 * @internal
+	 * @param LookupDataStore $data_store The data store to use.
+	 */
+	final public function init( LookupDataStore $data_store ) {
+		$this->data_store        = $data_store;
+		$this->lookup_table_name = $data_store->get_lookup_table_name();
+	}
+
+	/**
+	 * Checks if the product attribute filtering via lookup table feature is enabled.
+	 *
+	 * @return bool
+	 */
+	public function filtering_via_lookup_table_is_active() {
+		return 'yes' === get_option( 'woocommerce_attribute_lookup__enabled' );
+	}
+
+	/**
+	 * Adds post clauses for filtering via lookup table.
+	 * This method should be invoked within a 'posts_clauses' filter.
+	 *
+	 * @param array     $args Product query clauses as supplied to the 'posts_clauses' filter.
+	 * @param \WP_Query $wp_query Current product query as supplied to the 'posts_clauses' filter.
+	 * @param array     $attributes_to_filter_by Attribute filtering data as generated by WC_Query::get_layered_nav_chosen_attributes.
+	 * @return array The updated product query clauses.
+	 */
+	public function filter_by_attribute_post_clauses( array $args, \WP_Query $wp_query, array $attributes_to_filter_by ) {
+		global $wpdb;
+
+		if ( ! $wp_query->is_main_query() || ! $this->filtering_via_lookup_table_is_active() ) {
+			return $args;
+		}
+
+		$clause_root = " {$wpdb->prefix}posts.ID IN (";
+		if ( 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' ) ) {
+			$in_stock_clause = ' AND in_stock = 1';
+		} else {
+			$in_stock_clause = '';
+		}
+
+		foreach ( $attributes_to_filter_by as $taxonomy => $data ) {
+			$all_terms                  = get_terms( $taxonomy, array( 'hide_empty' => false ) );
+			$term_ids_by_slug           = wp_list_pluck( $all_terms, 'term_id', 'slug' );
+			$term_ids_to_filter_by      = array_values( array_intersect_key( $term_ids_by_slug, array_flip( $data['terms'] ) ) );
+			$term_ids_to_filter_by      = array_map(
+				function( $id ) {
+					return (int) $id;
+				},
+				$term_ids_to_filter_by
+			);
```

Comment:
Optional thing—passing `'absint'` instead of our own closure would be arguably more readable (at least, it would be shorter/would allow this to be a one-liner), plus I imagine we don't expect negative ints.

Reply:
I just wanted to prevent an unnecessary call to `abs`, but I guess the performance gain would be negligible compared with the readability gain so 👍🏻